stan/math/rev/mat/fun/cov_exp_quad.hpp

```diff
@@ -0,0 +1,305 @@
+         *    Assumes each element of x is the same size
+         * @param sigma standard deviation
+         * @param l length scale
+         */
+        cov_exp_quad_vari(const std::vector<T_x>& x,
+                          double sigma,
+                          const T_l& l)
+          : vari(0.0),
+          size_(x.size()),
+          size_vech_(size_ * (size_ - 1) / 2),
+          l_d_(value_of(l)), sigma_d_(value_of(sigma)),
+          sigma_sq_d_(std::pow(sigma_d_, 2)),
+          dist_(ChainableStack::memalloc_.alloc_array<double>(size_vech_)),
+          l_vari_(l.vi_),
+          cov_lower_(ChainableStack::memalloc_.alloc_array<vari*>(size_vech_)),
+          cov_upper_(ChainableStack::memalloc_.alloc_array<vari*>(size_vech_)),
+          cov_diag_(ChainableStack::memalloc_.alloc_array<vari*>(size_)) {
+            double inv_half_sq_l_d = 0.5 / (std::pow(l_d_, 2));
+            size_t pos = 0;
+            for (size_t j = 0; j < static_cast<size_t>(size_ - 1); ++j) {
+              for (size_t i = j + 1; i < static_cast<size_t>(size_); ++i) {
+                double dist_sq = squared_distance(x[i], x[j]);
+                double val = sigma_sq_d_ * exp(-dist_sq
+                                              * inv_half_sq_l_d);
+                dist_[pos] = dist_sq;
+                cov_upper_[pos] = new vari(val, false);
+                cov_lower_[pos] = new vari(val, false);
+                ++pos;
+              }
+            }
+              for (size_t i = 0; i < static_cast<size_t>(size_); ++i)
+                cov_diag_[i] = new vari(sigma_sq_d_, false);
+          }
+
+        virtual void chain() {
+          double adjl(0.0);
+
+          for (size_t i = 0; i < static_cast<size_t>(size_vech_); ++i) {
+            vari* el_low = cov_lower_[i];
+            vari* el_high = cov_upper_[i];
+            double prod_adj_cov_low = el_low->adj_ * el_low->val_;
+            double prod_adj_cov_high = el_high->adj_ * el_high->val_;
+            double prod_add = prod_adj_cov_low + prod_adj_cov_high;
+            adjl += prod_add * dist_[i];
+          }
+          l_vari_->adj_ +=  adjl / std::pow(l_d_, 3);
+        }
+    };
+
+    /**
+     * Returns a squared exponential kernel.
+     *
+     * @param x std::vector input that can be used in square distance
+     *    Assumes each element of x is the same size
+     * @param sigma standard deviation
+     * @param l length scale
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, or
+     *   x is nan or infinite
+     */
+    inline
+    Eigen::Matrix<var, -1, -1>
+    cov_exp_quad(const std::vector<double>& x,
+                 const var& sigma,
+                 const var& l) {
+      check_positive("cov_exp_quad", "sigma", sigma);
+      check_positive("cov_exp_quad", "l", l);
+      for (size_t n = 0; n < x.size(); n++)
+        check_not_nan("cov_exp_quad", "x", x[n]);
+
```

Comment:
I think there's a vector form of `check_not_nan` which will provide better error messages that include the index rather than just saying that `x` is not-a-number.